### DISCUSSION TITLE
Files path converting with convertFileSrc is incredibly hard

### DISCUSSION THREAD

**Discussion by erikpa1** [tauri-apps/tauri#5166] [Q&A]
Hi guys, 

as I mentioned in questions before, I'm creating 3D assets manager, where I use three js/fiber, where I load files asynchronous. Tauri is working pretty nice, but I found one big BUT.

Every possible HTTP path I have to convert via convertFileSrc() which is async and React fiber components crash where path is invalid or empty for example. To solve this I have to create parent wrapper, where I prepare path, and then load component, which is incredibly time consuming and making my code base bigger, and more mistakes happens,.. for example:

**How it should looks normaly on web:**
```
function _FrontSphere({panoramaPath, rotationY}) {
    const frontSphereTexture = useLoader(TextureLoader, panoramaPath)

    return (
        <mesh scale={[-1, 1, 1]}>
            <sphereGeometry args={[1020, 64, 64]}/>
            <meshBasicMaterial side={three.DoubleSide} map={frontSphereTexture}/>
        </mesh>
    )
}
```
But this magic I have to do because of tauri:

```
function FrontSphere({panoramaAsset, rotationY}) {

    const [path, setPath] = React.useState("")

    React.useEffect(() => {
        StorageManager.convertResourcesFileResourcePath(`${StorageManager.getAssetPrefix()}${panoramaAsset.getAbsolutePath()}`).then((path) => {
            setPath(path)
            console.log(path)
        })
    }, [])

    if (path === "") {
        return <></>
    } else {

  
      return <_FrontSphere panoramaPath={panoramaPath}/>
    }
}
```

As you can see there is almost 12 more lines of code, my code should have,... and this is only 1 asset. For those who don't know react: I have to create placeholder class which waits until tauri asynchronously calculate my file path and then instead of placeholder render real component.

Not sure what are solutions for this, but from my side, it would be nice to implement one of there:

1. public folder, where GET method is redirected - can be near .exe (or in config)
2. client side not asynchronous path converting ? something like convertFileSrc**Sync** ??

Or I'm I missing something ? :D

**Comment by vdvman1** [comment]
From what I can see, `convertFileSrc` already is synchronous, it returns a direct string:
- [Source code](https://github.com/tauri-apps/tauri/blob/2901145c497299f033ba7120af5f2e7ead16c75a/tooling/api/src/tauri.ts#L121)
- [Docs](https://tauri.app/v1/api/js/modules/tauri#convertfilesrc)

Where are you getting that it is an async function?

**Reply by erikpa1** [reply]
Oh, my stupid head,

I forgot to mention: 

```
    static async convertResourcesFileResourcePath(path) {

        const appDirPath = await resourceDir()
        const joinedPath = await join(appDirPath, path)
        const assetUrl = convertFileSrc(joinedPath);
        return assetUrl
    }
```

which is async because of resourceDir() ,...  sorry for missunderstanding


BUT, I found solution by myself already, by finding out convertFileSrc is sync, so just on app init remembered BaseDirectories.Resource + resourceDir(), and now I just do  convertFileSrc(RESOURCES + myPath).

Thanks for a little tip,  I figured the way :) now it works perfectly.

**Reply by vdvman1** [reply]
Glad you figured out a solution!
